### PR TITLE
Automatically generate protocol.d.ts by pulling in necessary dependencies

### PR DESCRIPTION
**Pull request by vladima**
As a result `protocol.d.ts` can be compiled as a standalone file (modulo `lib.d.ts`)
Also in this PR 
- added a few missing request/response types 
- assign string literal command types to requests
- hide VS specific command that should not be part of the protocol
- hide position argument on `FileLocationRequest` - we'll do work to properly split messages to line/offset-based and position-based later

Example of generated `protocol.d.ts`: https://gist.github.com/vladima/411ec6893d4e07a08c4804afbf0d754a

pending work: reflect changes made in `Jakefile.js` in `gulp` script
fixes: #11520, #11519, #11518, #11516, #11515, #11514. #11482, #11487

// CC: @mhegazy, @dbaeumer

### CODE REVIEW

**Comment by dbaeumer** [comment]
Looked at https://gist.github.com/vladima/411ec6893d4e07a08c4804afbf0d754a. Looks very nice.

**Comment by dbaeumer** [comment]
One minor comment: this is actually breaking names:

EditorOptions -> EditorSettings
FormatOptions -> FormatCodeSettings

OK with me but you might want to doc this.

File: src/server/protocol.ts [hunk 1]
```diff
@@ -1077,11 +1220,10 @@ declare namespace ts.server.protocol {
          * End position of the range for which to format text in file.
          */
         endPosition?: number;
```
Comment (mhegazy) [inline src/server/protocol.ts]: this should be internal too

File: Jakefile.js [hunk 2]
```diff
@@ -474,6 +472,40 @@ compileFile(processDiagnosticMessagesJs,
...
             /*useBuiltCompiler*/ false);
 
+var buildProtocolTs = path.join(scriptsDirectory, "buildProtocol.ts");
+var buildProtocolJs = path.join(scriptsDirectory, "buildProtocol.js");
+var buildProtocolDts = path.join(builtLocalDirectory, "protocol.d.ts");
+var typescriptServicesDts = path.join(builtLocalDirectory, "typescriptServices.d.ts");
+
+file(buildProtocolTs);
+
+compileFile(buildProtocolJs,
```
Comment (mhegazy) [inline Jakefile.js]: we need to add the new file to the LKG list

**Pull request merged**